### PR TITLE
[MRG] Make all plots with time end at tstop

### PR DESCRIPTION
**Pull request by tianqi-cheng**
Continues the PR #604.
Closes the issue #544.

### CODE REVIEW

**Comment by ntolley** [comment]
@rythorpe @jasmainak we decided to just cherry-pick from the old PR since it was paused so long ago (and we didn't have access to push to the PR directly)

File: hnn_core/dipole.py [hunk 1]
```diff
@@ -467,16 +467,11 @@ def savgol_filter(self, h_freq):
                                             self.sfreq)
         return self
 
-    def plot(self, tmin=None, tmax=None, layer='agg', decim=None, ax=None,
```
Comment (jasmainak) [inline hnn_core/dipole.py:470]: this is technically a backwards incompatible change. If some user has `tmin` and `tmax` in their script, it will stop working now. The correct fix is to add a deprecation cycle. But if we want to assume that it's unlikely many users have this in their scripts, we should mark it as a "Bugfix" in `whats_new.rst` ...

Comment (jasmainak) [reply hnn_core/dipole.py:470]: I thought one reason to have `tmin` was for burn in period @rythorpe may have an opinion ...

but it's true we can probably just set `plt.xlim((tmin, tmax))` to achieve the same

Comment (rythorpe) [reply hnn_core/dipole.py:470]: See my comment in the previous PR [here](https://github.com/jonescompneurolab/hnn-core/pull/604#issuecomment-1462770587). I don't think we need `tmin` at the plotting level, however, it would be useful at the `simulate_dipole` level for a burn-in period. That's a very different feature though and should probably be implemented in a separate PR.

Comment (rythorpe) [reply hnn_core/dipole.py:470]: I agree that we should do a deprecation cycle for these input arguments though.

File: hnn_core/viz.py [hunk 2]
```diff
@@ -182,6 +178,7 @@ def plot_laminar_lfp(times, data, contact_labels, tmin=None, tmax=None,
             col = color
         ax.plot(plot_times, plot_data + trace_offsets[contact_no],
                 label=f'C{contact_no}', color=col)
+        ax.set_xlim(right=plot_times[-1])
```
Comment (rythorpe) [inline hnn_core/viz.py:181]:
```suggestion
        ax.set_xlim(right=times[-1])
```
`xlim` should always end at the same point regardless of whether or not the time series was decimated.

File: hnn_core/viz.py [hunk 3]
```diff
@@ -491,6 +486,10 @@ def plot_spikes_raster(cell_response, trial_idx=None, ax=None, show=True):
         The CellResponse object from net.cell_response
     trial_idx : int | list of int | None
         Index of trials to be plotted. If None, all trials plotted
+    tmin : float or None
+        Start time of plot in milliseconds. If None, plot entire simulation.
+    tmax : float or None
+        End time of plot in milliseconds. If None, plot entire simulation.
```
Comment (rythorpe) [inline hnn_core/viz.py:492]: Why was this added? I agree that time bounds should be included somehow when generating a raster plot, but I believe we can just use `cell_response.times`.

File: hnn_core/viz.py [hunk 4]
```diff
@@ -1189,6 +1189,9 @@ def plot_laminar_csd(times, data, contact_labels, ax=None, colorbar=True,
         Sampling times (in ms).
     data : array-like, shape (n_channels, n_times)
         CSD data, channels x time.
+    tmin : float | None
+        Start time of plot in milliseconds. If None, plot entire simulation.
+    tmax : float | None
```
Comment (rythorpe) [inline hnn_core/viz.py:1194]: Not sure about this either. Is it necessary?

File: hnn_core/viz.py (hunk 3 above)
Comment (ntolley) [reply hnn_core/viz.py:492]: @rythorpe for reference a lot of these were just copied from the previous PR. @tianqi-cheng you can remove tmin/tmax

Comment (rythorpe) [reply hnn_core/viz.py:492]: Ah okay. Since the diff is different here than in #604, I guess some of commits didn't get copied over. Perhaps we should cherry-pick the remaining commits so that we don't have to rehash some of our previous discussion points?

Comment (tianqi-cheng) [reply hnn_core/viz.py]: We did cherry-pick, but PR had open comments that were not addressed. I just updated the codes and deleted the tmin and tmax and other related codes.

File: hnn_core/viz.py (hunk 4 above)
Comment (tianqi-cheng) [reply hnn_core/viz.py]: Also deleted them too.

**Comment by tianqi-cheng** [comment]
@rythorpe Hi Ryan, do you think I addressed the issues properly?

File: hnn_core/viz.py [hunk 5]
```diff
@@ -288,19 +281,19 @@ def plot_dipole(dpl, tmin=None, tmax=None, ax=None, layer='agg', decim=None,
...
                 # extract scaled data and times
-                data, times = _get_plot_data_trange(dpl_trial.times,
-                                                    dpl_trial.data[layer],
-                                                    tmin, tmax)
+                data = dpl_trial.data[layer]
+                times = dpl_trial.times
                 if decim is not None:
-                    data, times = _decimate_plot_data(decim, data, times)
+                    data, times = _decimate_plot_data(
+                        decim, dpl_trial.data[layer], dpl_trial.times)
```
Comment (rythorpe) [inline hnn_core/viz.py:288]:
```suggestion
                    data, times = _decimate_plot_data(decim, data, times)
```
Just to simplify...

File: hnn_core/viz.py [hunk 6]
```diff
@@ -538,6 +533,9 @@ def plot_spikes_raster(cell_response, trial_idx=None, ax=None, show=True):
         cell_type_times, cell_type_ypos = [], []
         for gid in cell_type_gids:
             gid_time = spike_times[spike_gids == gid]
+            _, gid_time = _get_plot_data_trange(
+                gid_time, gid_time, cell_response.times[0],
+                cell_response.times[-1])
```
Comment (rythorpe) [inline hnn_core/viz.py:538]: Is there a reason why we need this? I think any spike times outside the x-limits will get cropped when plotting.

File: hnn_core/viz.py [hunk 7]
```diff
@@ -1218,6 +1218,7 @@ def plot_laminar_csd(times, data, contact_labels, ax=None, colorbar=True,
 
     ax.set_xlabel('Time (ms)')
     ax.set_ylabel('Electrode depth')
+    ax.set_xlim(right=times[-1])
```
Comment (rythorpe) [inline hnn_core/viz.py:1221]: This is a nitpick, but I'd recommend defining `tstop` (or whatever you want to call it) before decimating so that you don't introduce a small error in the right x-limit.

**Review by rythorpe — COMMENTED** [review]
This looks good @tianqi-cheng. The next steps will be to resolve my minor comments below and add a deprecation warning + updating `whats_new.rst` based on @jasmainak's recommendation.

One other issue I anticipate we'll need to address before merging is that `CellResponse` instances read in from spike files don't set the `CellResponse.times` attribute. Since the code below relies on this attribute always being defined, we'll need to find a fix for this.

**Comment by tianqi-cheng** [comment]
> This looks good @tianqi-cheng. The next steps will be to resolve my minor comments below and add a deprecation warning + updating `whats_new.rst` based on @jasmainak's recommendation.
> 
> One other issue I anticipate we'll need to address before merging is that `CellResponse` instances read in from spike files don't set the `CellResponse.times` attribute. Since the code below relies on this attribute always being defined, we'll need to find a fix for this.

Thanks! I will work on it.

File: hnn_core/viz.py (hunk 7 above)
Comment (rythorpe) [reply hnn_core/viz.py:1220]: Ah you're right. I didn't realize that [`_get_plot_data_trange`](https://github.com/tianqi-cheng/hnn-core/blob/d2c5c8c506dcbde33856f0fa3519674ab4a391b3/hnn_core/viz.py#L24) only crops the time series. Since this is the case, however, I don't think we need the added lines on L1207-1209 as they crop the time series to its original length (i.e., they don't actually crop anything off).

Comment (rythorpe) [reply hnn_core/viz.py:1220]: Re: adding the decimate feature to `plot_laminar_csd`, my vote would be to save it for a separate PR.

**Comment by katduecker** [comment]
Hey, I have now taken on this pull request. I'm thinking about the options for the deprecation cycle of defining tmin and tmax for plot_dipole et al. 
1. Should it still be possible to set the xlims using tmax and tmin, and if tmax and tmin aren't defined then rasterplots and dipole are plotted from 0 to stop?
2. OR should the data always be plotted from 0 to tstop, and the deprecation warning tells the user to set xlim after plotting? (i.e. setting tmin and tmax doesn't do anything) 
I'm happy to make all of these clear in the tutorials when I'm done!
